module/zfs/zio_compress.c

```diff
@@ -47,24 +47,42 @@ unsigned long zio_decompress_fail_fraction = 0;
  * Compression vectors.
  */
 zio_compress_info_t zio_compress_table[ZIO_COMPRESS_FUNCTIONS] = {
-	{"inherit",		0,	NULL,		NULL},
-	{"on",			0,	NULL,		NULL},
-	{"uncompressed",	0,	NULL,		NULL},
-	{"lzjb",		0,	lzjb_compress,	lzjb_decompress},
-	{"empty",		0,	NULL,		NULL},
-	{"gzip-1",		1,	gzip_compress,	gzip_decompress},
-	{"gzip-2",		2,	gzip_compress,	gzip_decompress},
-	{"gzip-3",		3,	gzip_compress,	gzip_decompress},
-	{"gzip-4",		4,	gzip_compress,	gzip_decompress},
-	{"gzip-5",		5,	gzip_compress,	gzip_decompress},
-	{"gzip-6",		6,	gzip_compress,	gzip_decompress},
-	{"gzip-7",		7,	gzip_compress,	gzip_decompress},
-	{"gzip-8",		8,	gzip_compress,	gzip_decompress},
-	{"gzip-9",		9,	gzip_compress,	gzip_decompress},
-	{"zle",			64,	zle_compress,	zle_decompress},
-	{"lz4",			0,	lz4_compress_zfs, lz4_decompress_zfs}
+	{"inherit",		0,	NULL,		NULL, NULL},
+	{"on",			0,	NULL,		NULL, NULL},
+	{"uncompressed",	0,	NULL,		NULL, NULL},
+	{"lzjb",		0,	lzjb_compress,	lzjb_decompress, NULL},
+	{"empty",		0,	NULL,		NULL, NULL},
+	{"gzip-1",		1,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-2",		2,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-3",		3,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-4",		4,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-5",		5,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-6",		6,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-7",		7,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-8",		8,	gzip_compress,	gzip_decompress, NULL},
+	{"gzip-9",		9,	gzip_compress,	gzip_decompress, NULL},
+	{"zle",			64,	zle_compress,	zle_decompress, NULL},
+	{"lz4",		0,	lz4_compress_zfs, lz4_decompress_zfs, NULL},
+	{"zstd",	0,	zstd_compress, zstd_decompress, zstd_getlevel}
 };
 
+enum zio_zstd_levels
+zio_zstd_level_select(spa_t *spa, enum zio_zstd_levels child,
+    enum zio_zstd_levels parent)
+{
+	enum zio_zstd_levels result;
+
+	ASSERT(child < ZIO_ZSTDLVL_LEVELS);
+	ASSERT(parent < ZIO_ZSTDLVL_LEVELS);
+	ASSERT(parent != ZIO_ZSTDLVL_INHERIT);
```

Comment:
nit: `ASSERT3U` and `ASSERTP` here.

Reply:
ASSERTP does not exist, but ASSERT3P does. but ASSERT3P makes no sense since its for pointers, but these values are no pointers, just enums. its also curious why we should cast enums to uint64 which is waste of space and will decrease performance in 32 bit systems (okay its assert and only available in debug mode)

Reply:
The advantage to using the `ASSERT3*` macros is that the exact values that caused the failure will be included in the ASSERT console message and consequently system logs.  Having the values and the stack trace is often enough information alone to debug the issue.

Reply:
your request is included in latest revision. but consider that alot of other parts in zio_compress unrelated to my PR are still using the old ASSERT macro.